rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java

```diff
@@ -0,0 +1,189 @@
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOG = LoggerFactory.getLogger(RocketMQSink.class);
+
+    private transient DefaultMQProducer producer;
+    private boolean async; // false by default
+
+    private Properties props;
+    private TopicSelector<IN> topicSelector;
+    private KeyValueSerializationSchema<IN> serializationSchema;
+
+    private boolean batchFlushOnCheckpoint; // false by default
+    private List<Message> batchList;
+
+    public RocketMQSink(KeyValueSerializationSchema<IN> schema, TopicSelector<IN> topicSelector, Properties props) {
+        this.serializationSchema = schema;
+        this.topicSelector = topicSelector;
+        this.props = props;
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        Validate.notEmpty(props, "Producer properties can not be empty");
+        Validate.notNull(topicSelector, "TopicSelector can not be null");
+        Validate.notNull(serializationSchema, "KeyValueSerializationSchema can not be null");
+
+        producer = new DefaultMQProducer();
+        producer.setInstanceName(String.valueOf(getRuntimeContext().getIndexOfThisSubtask()));
+        RocketMqConfig.buildProducerConfigs(props, producer);
+
+        batchList = new LinkedList<>();
+
+        if (batchFlushOnCheckpoint && !((StreamingRuntimeContext) getRuntimeContext()).isCheckpointingEnabled()) {
+            LOG.warn("Flushing on checkpoint is enabled, but checkpointing is not enabled. Disabling flushing.");
+            batchFlushOnCheckpoint = false;
+        }
+
+        try {
+            producer.start();
+        } catch (MQClientException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void invoke(IN input, Context context) throws Exception {
+        Message msg = prepareMessage(input);
+
+        if (batchFlushOnCheckpoint) {
+            batchList.add(msg);
+            return;
+        }
+
+        if (async) {
+            // async sending
+            try {
+                producer.send(msg, new SendCallback() {
+                    @Override
+                    public void onSuccess(SendResult sendResult) {
+                        LOG.debug("Async send message success!");
+                    }
+
+                    @Override
+                    public void onException(Throwable throwable) {
+                        if (throwable != null) {
+                            LOG.error("Async send message failure!", throwable);
+                        }
+                    }
+                });
+            } catch (Exception e) {
+                LOG.error("Async send message failure!", e);
+            }
+        } else {
+            // sync sending, will return a SendResult
+            try {
+                SendResult result = producer.send(msg);
+                if (LOG.isDebugEnabled()) {
```

Comment:
No need to check whether debug is enabled.

Reply:
Yes. Will remove.